### PR TITLE
Allow to specify file encoding

### LINKED ISSUE #473: Allow to specify file encoding
Currently the file encoding is always the system default encoding. This is unreliable. It would be better to have a way to specify the encoding of the generated .java files.

### PR DESCRIPTION
**Pull request by marschall**
Add the ability to specify the file encoding.

Closes #473

### CODE REVIEW

**Comment by marschall** [comment]
I agree to and have signed the CLA.

**Comment by swankjesse** [comment]
When you customize the charset, is it to force UTF-8? Or is there another charset you’re using? I’m specifically wondering whether we should just change the defaults to always write UTF-8, since that’s the only good charset there is.

**Comment by tbroyer** [comment]
Well, in an annotation processor, at least, JavaPoet should use the default encoding/charger, otherwise it could break the compilation. We can all agree that everyone should be using UTF-8, but this shouldn't be enforced (at least in this case).

File: src/main/java/com/squareup/javapoet/JavaFile.java [hunk 1]
```diff
@@ -94,7 +102,8 @@ public void writeTo(Path directory) throws IOException {
     }
 
     Path outputPath = outputDirectory.resolve(typeSpec.name + ".java");
-    try (Writer writer = new OutputStreamWriter(Files.newOutputStream(outputPath))) {
+    try (Writer writer = Files.newBufferedWriter(outputPath,
+            charset != null ? charset : Charset.defaultCharset())) {
```
Comment (tbroyer) [inline src/main/java/com/squareup/javapoet/JavaFile.java]: Personal opinion: I'd rather have it throw if `null`, and have the overload explicitly pass the default charset in.

**Comment by sormuras** [comment]
Imho, UTF-8 should be the default in JavaPoet.

Even if the encoding was set when writing, the annotation processor calling environment might disregard that and use whatever it likes: https://bugs.eclipse.org/bugs/show_bug.cgi?id=483690

**Comment by tbroyer** [comment]
OK, disregard my comment, `writeTo(Filer)` uses `JavaFileObject#openWriter`. OK to default to UTF-8 for the `File` and `Path` overloads then; anyone wanting a different encoding could still use the `Appendable` overload. Sorry for the noise.

**Comment by swankjesse** [comment]
Does anyone object to making JavaPoet force UTF-8, without the option for further configuration? As far as I know that should make everyone happy.

**Comment by marschall** [comment]
Some of our projects are on UTF-8. Others are on ISO-8859-1. In this specific case it was to force UTF-8 because in Windows (:cry:) the default encoding was CP-1252.

AFAIK Eclipse defaults whatever the platform encoding is (see [bug 108668](https://bugs.eclipse.org/bugs/show_bug.cgi?id=108668)) so hardcoding UTF-8 would cause more configuration effort for Windows Eclipse users.

IMHO the ideal solution would be if JavaPoet generated [unicode escapes](https://docs.oracle.com/javase/specs/jls/se7/html/jls-3.html#jls-3.3) for everything non-ASCII. That avoids all encoding issues (unless you use EBCDIC).

File: src/main/java/com/squareup/javapoet/JavaFile.java (hunk 1 above)
Comment (marschall) [inline src/main/java/com/squareup/javapoet/JavaFile.java]: fixed

**Comment by swankjesse** [comment]
I’m still pretty undecided on this. Forcing UTF-8 seems reasonable, and forcing the platform’s default charset also seems reasonable. Would we ever want a different charset than those two? Right now we use the platform’s default charset.

**Comment by swankjesse** [comment]
I’m forcing UTF-8.
https://github.com/square/javapoet/pull/522

**Pull request closed without merging**